docs/design/declaring_entities.md

```diff
@@ -0,0 +1,233 @@
+# Declaring entities
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Overview](#overview)
+-   [Matching redeclarations of an entity](#matching-redeclarations-of-an-entity)
+    -   [Details](#details)
+-   [`extern` and `extern library`](#extern-and-extern-library)
+    -   [Effect on indirect imports](#effect-on-indirect-imports)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Overview
+
+Entities may have up to three declarations:
+
+-   An optional, owning forward declaration.
+    -   For example, `class MyClass;`.
+    -   This must come before the definition. The API file is considered to be
+        before the implementation file.
+-   A required, owning definition.
+    -   For example, `class MyClass { ... }`.
+    -   The definition might be the _only_ declaration.
+-   An optional, non-owning `extern library "<owning_library>"` declaration.
+    -   For example, `extern library "OtherLibrary" class MyClass;`.
+    -   It must be in a separate library from the definition.
+    -   The owning library's API file must import the `extern` declaration, and
+        must also contain a declaration.
+    -   The owning library's declarations must have the `extern` modifier
+        (without `library`).
+        -   For example, `extern class MyClass;`.
+
+For example, a library can have a forward declaration of an entity in the API
+file, and use the implementation file for the entity's definition. Putting the
+definition in an implementation file this way can reduce the dependencies for
+API file evaluation, improving compile time. This is commonly done with
+functions. For example:
+
+```
+library "MyLibrary";
+
+fn DoSomething();
+```
+
+```
+impl library "MyLibrary";
+
+fn DoSomething {
+  ...
+}
+```
+
+## Matching redeclarations of an entity
+
+In order to determine whether two redeclarations refer to the same entity, we
+apply the rules:
+
+-   Two named declarations _declare the same entity_ if they have the same scope
+    and the same name.
+-   One declaration _redeclares_ another if they declare the same entity and the
+    second declaration appears after the first, including the case where the
+    first declaration is imported. In this case, the second declaration is said
+    to be a _redeclaration_ of the first.
```

Comment:
This seems longer than necessary. How about

```suggestion
-   When two named declarations declare the same entity, the second is said to be a
    _redeclaration_, including the case where the first declaration is imported.
```

Is there something else that this doesn't capture that you were trying to convey?

Reply:
To be sure, this was directly copied from #3763, but I think what you're saying is getting at what zygoloid was trying to say. I'm moving the comment about imports to the first bullet though, since it's not specific to the first declaration.